Allow \" in single-double-quoted string interpolations

eed3si9n opened the pull request:
Changing `"Hello, \"World\""` to `s"Hello, \"$who\""` no longer breaks.

Before this change, `\"` terminated single-quoted interpolated string
literals, now the string remains open. The scanner doesn't interpret
the escape sequence, string interpolators can do so (`s` and `f` do).

Breaking changes:
  - `raw"c:\"` no longer compiles, it's now an unclosed string
  - `raw"c:\" // uh"` used to evaluate to `"""c:\"""`, now it's `"""c:\" // uh"""`

Fixes https://github.com/scala/bug/issues/6476

eed3si9n committed:
Allow \" in single-quoted string interpolations

Changing `"Hello, \"World\""` to `s"Hello, \"$who\""` no longer breaks.

Before this change, `\"` terminated single-quoted interpolated string
literals, now the string remains open. The scanner doesn't interpret
the escape sequence, string interpolators can do so (`s` and `f` do).

Breaking changes:
  - `raw"c:\"` no longer compiles, it's now an unclosed string
  - `raw"c:\" // uh"` used to evaluate to `"""c:\"""`, now it's
    `"""c:\" // uh"""`

SethTisue commented:
> This change should be relatively safe

I think we need to actually know what the safety level is. Either it's absolutely safe, or there are unsafe cases that need to be precisely characterized in order to know whether we're comfortable with advancing this.

eed3si9n commented:
> This change should be relatively safe since neither `s"\"foo\""` nor `s"foo\"` previously compiled.

One test failure in [Jenkins](https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/11400/console) was a negative test.

```diff
% diff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/neg/t8266-invalid-interp.check /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/neg/t8266-invalid-interp-neg.log
@@ -1,10 +1,4 @@
-t8266-invalid-interp.scala:4: error: Trailing '\' escapes nothing.
+t8266-invalid-interp.scala:4: error: unclosed string literal
    ​f"a\",
-       ^
-t8266-invalid-interp.scala:5: error: invalid escape '\x' not one of [\b, \t, \n, \f, \r, \\, \", \', \uxxxx] at index 1 in "a\xc". Use \\ for literal \.
-    f"a\xc",
-       ^
-t8266-invalid-interp.scala:7: error: invalid escape '\v' not one of [\b, \t, \n, \f, \r, \\, \", \', \uxxxx] at index 1 in "a\vc". Use \\ for literal \.
-    f"a\vc"
-       ^
-3 errors
+      ^
+1 error
```

So the code that previously did not compile gets a different error message.

### breaking change

One breaking change I can think of is `raw` or any other custom interpolator with a String part ending in `\`. Like `raw"foo\"`.

```scala
foo.scala:9: error: unclosed string literal
     val raw2 = raw"foo\"
                    ^
```

In these cases the workaround would be to migrate to `raw"""..."""`.

som-snytt commented:
Not sure if you reviewed https://github.com/scala/scala/pulls?q=is%3Apr+author%3Asom-snytt+6476+is%3Aclosed

especially https://github.com/scala/scala/pull/4308 has a dreary tail.

eed3si9n commented:
@som-snytt I did see some of the discussion that took place, and that's one of my motivation behind lawyering up on the legality within current statute.

eed3si9n commented:
I cherry picked https://github.com/scala/scala/pull/4308/commits/195d3699b8d2fced71fb8fd4fb38eddf746f2935 from Som's 2015 PR. This handles last backslash, so it's better than my implementation.

OlegYch commented:
all escapes should be treated in the same way, not just \"

eed3si9n commented:
@SethTisue Let's define safety as non-existence of:
- semantic changes to previously-compiled binaries
- semantic changes to previously-admissible sources
- admittance of illegal program

We can rule out first and last I think. Is there a source that we can write that would change the meaning? `raw`-interpolator captures the character sequence, so if there's a problem, we should see it using `raw`.

- My first commit didn't handle backslash in last position `raw"foo\"`. That's now fixed with Som's change.
- That in turn introduced `\$` changing the semantics of `raw"\$foo"`. I rolled that back.

However I _can_ make up a program that would evaluate differently:

```scala
println(raw"1\"); println(2); // \"")
```

In Scala 2.13.1 this would print:

```scala
scala> println(raw"1\"); println(2); // \"")
1\
2
```

With this patch:

```scala
1\"); println(2); // \"
```

It would be very unlikely that you'd have a comment `// \"")` lying around there so it would more likely fail to compile.

som-snytt commented:
I also noticed that looking ahead for another quote doesn't handle comments; it's a pretty severe limitation, but only worth fixing if the PR has a chance.

eed3si9n commented:
@OlegYch 

> all escapes should be treated in the same way, not just \" 

The other backslash escape sequences like `\n` already work with `s"..."`. It's just that `\"` currently is treated as the-end-of-literal by the scanner, which this PR aims to fix. Do you have specific example that I missed?

OlegYch commented:
@eed3si9n sorry i thought this was related to raw strings, nvm then

dwijnand commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -887,70 +887,89 @@ trait Scanners extends ScannersCommon {
-        } else {
+        case '"' if escaped =>
+          // at \", unescape and finish if this is the last quote on the line
+          if (!multiLine && isLastQuoteOnLine) {
```
`multiLine` can only be false here, so `!multiLine` can be dropped.

dwijnand replied:
Let's move this into the guard, so we don't duplicate the RHS of the next case.

dwijnand reviewed (commented):
Some comments from reviewing.

dwijnand commented:
Also, we should fix the REPL, as it behaves differently to real compilation:

```scala
scala> raw"\\"
val res0: String = \
scala> raw"\"
val res1: String = ""
scala> "\""
     |
     |
You typed two blank lines.  Starting a new command.
scala> "\\"
       ^
       error: unclosed string literal
```

lrytz commented:
> It would be very unlikely that you'd have a comment `// \"")` lying around there so it would more likely fail to compile.

Dale and me were trying to come up with a few more examples around this; so your point is that the change can only prevent code from compiling that didn't work before, or vice-versa, but not make compile and run code differently (except for the very corner case you highlight)?


this no longer compiles

```
$> spr 8830 -e 'print(raw"\"); println("")'
/var/folders/py/8q6hn5qj3g365g0tsy63f68c0000gn/T/scalacmd9448771494255440057.scala:1: error: unclosed string literal
print(raw"\"); println("")
                        ^
```

this now compiles

```
spr 8830 -e 'print(raw"\"); println(\"")'
\"); println(\"
```

I personally have doubts about speccing the `isLastQuoteOnLine` part. I think it's preferrable to go with the breaking change and treat `\"` in an interpolated string as not closing the string.

martijnhoekstra commented:
`$"` still makes more sense to me than `\"` as escaped quote: it breaks nothing, and it's more regular in that `$` is the meta-character for interpolations, while how `\` is treated is up to the interpolator itself: that it's treated the same as in regular strings in the s and f interpolator is a property of those interpolators, not of interpolations in general.

lrytz commented:
I agree with that in principle, and I think we should have the fix you did in Scala 3 also in Scala 2.

However, I really think for user-friendlyness moving from `"he said \"hi\""` to `s"$he said \"hi\""` should work. If it can be done without breaking too much.

dwijnand commented:
> `$"` still makes more sense to me than `\"` as escaped quote

A subtlety (perhaps you knew this already): this doesn't treat `\"` as an escaped quote, it treating it as _not_ the terminator of the string.  Compare how `raw` keeps `\"` but doesn't keep `$"`:

```scala
$ scala --scala-pr 8830 -e 'println(s"he said: \"hi\" to me")'
he said: "hi" to me
$ scala --scala-pr 8830 -e 'println(raw"he said: \"hi\" to me")'
he said: \"hi\" to me
$
$ scala3
Starting scala3 REPL...
scala> s"he said: $"hi$" to me"
val res0: String = he said: "hi" to me

scala> raw"he said: $"hi$" to me"
val res1: String = he said: "hi" to me
```

eed3si9n commented:
> I personally have doubts about speccing the `isLastQuoteOnLine` part. I think it's preferable to go with the breaking change and treat `\"` in an interpolated string as not closing the string.

If I remember correctly, I cherry picked the last-`\` support from a Som PR, so if we go with just the first commit in this PR that's likely what you want.

som-snytt commented:
https://github.com/scala/scala/pull/4308#issuecomment-86192223 is still true.

The old PR is still an easy rebase on 2.12:
```
scala> raw"C:\Users\snytt\"
res4: String = C:\Users\snytt\

scala> raw"C:\Users\snytt\"    with a stray trailing quote    "
res5: String = "C:\Users\snytt\"    with a stray trailing quote    "
```
People misplace quotes and braces all the time and experience brain failure until they figure it out.

A close quote is the next unescaped quote, or the last escaped quote on the line.

Many of the "what about this" examples don't pass the Odersky test of sane people don't code like that.

eed3si9n commented:
Rebased to 2.13.x and dropped the additional commits.

lrytz commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -910,6 +910,20 @@ trait Scanners extends ScannersCommon {
           setStrVal()
           token = STRINGLIT
         }
+      } else if (ch == '\\') {
```
```suggestion
      } else if (ch == '\\' && !multiLine) {
```
then we can drop the `if` below

lrytz commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -910,6 +910,20 @@ trait Scanners extends ScannersCommon {
+        } else {
+          putChar(ch)
+          nextRawChar()
+          if (ch == '"') {
```
```suggestion
          if (ch == '"' || ch == '\\') {
```

to support `s"\\"`?

dwijnand replied:
Aka #9536.

lrytz commented:
(t7020.scala is flaky currently, likely since https://github.com/scala/scala/pull/9522)

dwijnand commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -945,13 +960,14 @@ trait Scanners extends ScannersCommon {
-            unclosedStringLit()
+          else {
+            unclosedStringLit(seenEscapedQuote)
+          }
```
> Also, we should fix the REPL, as it behaves differently to real compilation:

Can we add a REPL test for this?

lrytz replied:
added

lrytz committed:
Spec for \" in interpolated strings

Also, unicode escapes are no longer interpreted in interpolated strings.
Interpolators can still interpret them, but that's not in the spec.

lrytz commented:
Pushed some spec'ese words for this change.

som-snytt commented on spec/01-lexical-syntax.md:
````diff
@@ -513,21 +514,23 @@ alphaid                ::= upper idrest
 ```
 
-Interpolated string consist of an identifier starting with a letter immediately 
+An interpolated string consist of an identifier starting with a letter immediately
````
consists

som-snytt commented on spec/01-lexical-syntax.md:
```diff
@@ -513,21 +514,23 @@ alphaid                ::= upper idrest
+multi-line (enclosed in triple quotes). Note that the sequence `\"` does not
+close a normal string literal (enclosed in single quotes).
+
+There are two new forms of dollar sign escape. 
```
new?

lrytz commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -854,7 +854,12 @@ trait Scanners extends ScannersCommon {
-    private def unclosedStringLit(): Unit = syntaxError("unclosed string literal")
+    private def unclosedStringLit(seenEscapedQuoteInInterpolation: Boolean = false): Unit = {
+      val note =
+        if (seenEscapedQuoteInInterpolation) "; note that `\\\"` no longer closes single-quoted interpolated string literals since 2.13.6, you can use a triple-quoted string instead"
```
This suggests to move from `s"..."` to `s"""..."""`. The two behave the same, because escapes are handled by `s` in both cases.

To me the main goal of this PR is to avoid the surprise of going from `"she: \"me?\""` to `s"she: \"me?\""`. The discrepancy between `"""yes\no"""` and `s"""yes\no"""` is much less sublte and should maybe also be addressed somehow...

dwijnand commented:
> This is now ready, but let's see what the Scala 3 people say over at [lampepfl/dotty#11751](https://github.com/lampepfl/dotty/pull/11751).

That was merged.  So shall we (f-i-nally!) merge this now?

SethTisue commented:
@som-snytt you'll have to stop saying "when string interpolators support escaped double quotes" and go back to saying "when hell freezes over" like before

@eed3si9n thanks for not giving up on this :-)

som-snytt commented:
I was going to go with flying pigs. My favorite part re-reading the ticket is when SethTisue says this is impossible now and eed3si9n submits a PR.

SethTisue commented:
>  My favorite part re-reading the ticket is when SethTisue says this is impossible now and eed3si9n submits a PR.

Can't wait to find out who plays me in the movie version.
